src/main/java/io/lakefs/PathMapper.java

```diff
@@ -0,0 +1,311 @@
+
+import java.io.IOException;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * A class that does path mapping between configured pairs of path styles, and matches the appropriate path mapping for
+ * a converted path.
+ */
+public class PathMapper {
+
+    public static final Logger LOG = LoggerFactory.getLogger(PathMapper.class);
+
+    protected static final String MAPPING_CONFIG_PREFIX = "routerfs.mapping.";
+    protected static final String NON_DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.([a-z0-9-_]*)\\.(\\d*)\\.(replace|with)";
+    protected static final String DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.(.*default)\\.(replace|with)";
+    protected static final int NON_DEFAULT_TO_FS_IDX = 1;
+    protected static final int NON_DEFAULT_MAPPING_IDX_IDX = 2;
+    protected static final int NON_DEFAULT_MAPPING_TYPE_IDX = 3;
+    protected static final int DEFAULT_TO_FS_IDX = 1;
+    protected static final int DEFAULT_MAPPING_TYPE_IDX = 2;
+
+    enum MappingConfigType {
+        REPLACE,
+        WITH
+    }
+
+    private List<PathMapping> pathMappings;
+    private PathMapping defaultMapping;
+
+    public PathMapper(Configuration conf) throws IOException {
+        this.pathMappings = new ArrayList<>();
+        loadMappingConfig(conf);
+        if (defaultMapping == null) {
+            throw new IllegalArgumentException("Missing default mapping configuration, cannot initialize path mapper");
+        }
+    }
+
+    private void loadMappingConfig(Configuration conf) throws InvalidPropertiesFormatException {
+        List<MappingConfig> mappingConfiguration = parseMappingConfig(conf);
+        populatePathMappings(mappingConfiguration);
+
+        if (LOG.isDebugEnabled()) {
+            logLoadedMappings();
+        }
+    }
+
+    /**
+     * Pair mapping configuration that form a {@link PathMapping}, and populate the pathMapping list.
+     * Two mapping configuration are paired together in case they share the same toFsScheme and mappingIdx, but have
+     * different {@link MappingConfigType}.
+     * e.g. the following two mapping configuration for a path mapping:
+     * routerfs.mapping.lakefs.1.replace='^s3a://bucket/prefix'
+     * routerfs.mapping.lakefs.1.with='lakefs://example-repo/dev/prefix'
+     *
+     * @param mappingConfiguration the configurations to create path mapping from
+     */
+    private void populatePathMappings(List<MappingConfig> mappingConfiguration) {
+        List<MappingConfig> srcStyles = mappingConfiguration.stream()
+                .filter(mc -> mc.getType() == MappingConfigType.REPLACE).collect(Collectors.toList());
+
+        for (MappingConfig srcStyle : srcStyles) {
+            Optional<MappingConfig> matchingDstStyle = mappingConfiguration.stream()
+                    .filter(mc -> mc.getToScheme().equals(srcStyle.getToScheme()) && mc.getIndex() == srcStyle.getIndex()
```

Comment:
This looks like a quadratic algorithm.  But given that everything matching has an index, couldn't this just be a linear algorithm that sets up a single data structure -- it just needs to populate a map of lists from a list, or something of the sort.

Reply:
done